App/Services/CredentialManager.cs

```diff
@@ -0,0 +1,247 @@
+            ApiToken = apiToken,
+        });
+    }
+
+    public void ClearCredentials()
+    {
+        NativeApi.DeleteCredentials(CredentialsTargetName);
+        UpdateState(new CredentialModel
+        {
+            State = CredentialState.Invalid,
+            CoderUrl = null,
+            ApiToken = null,
+        });
+    }
+
+    private void UpdateState(CredentialModel newModel)
+    {
+        _latestCredentials = newModel;
+        CredentialsChanged?.Invoke(this, _latestCredentials.Clone());
+    }
+
+    private RawCredentials? ReadCredentials()
+    {
+        var raw = NativeApi.ReadCredentials(CredentialsTargetName);
+        if (raw == null) return null;
+
+        RawCredentials? credentials;
+        try
+        {
+            credentials = JsonSerializer.Deserialize<RawCredentials>(raw);
+        }
+        catch (JsonException)
+        {
+            return null;
+        }
+
+        if (credentials is null || string.IsNullOrWhiteSpace(credentials.CoderUrl) ||
+            string.IsNullOrWhiteSpace(credentials.ApiToken)) return null;
+
+        return credentials;
+    }
+
+    private void WriteCredentials(RawCredentials credentials)
+    {
+        var raw = JsonSerializer.Serialize(credentials);
+        NativeApi.WriteCredentials(CredentialsTargetName, raw);
+    }
+
+    private class RawCredentials
+    {
+        public required string CoderUrl { get; set; }
+        public required string ApiToken { get; set; }
+    }
+
+    private static class NativeApi
+    {
+        private const int CredentialTypeGeneric = 1;
+        private const int PersistenceTypeLocalComputer = 2;
+        private const int ErrorNotFound = 1168;
+
+        public static string? ReadCredentials(string targetName)
+        {
+            if (!CredReadW(targetName, CredentialTypeGeneric, 0, out var credentialPtr))
+            {
+                var error = Marshal.GetLastWin32Error();
+                if (error == ErrorNotFound) return null;
+                throw new InvalidOperationException($"Failed to read credentials (Error {error})");
+            }
+
+            try
+            {
+                var cred = Marshal.PtrToStructure<CREDENTIAL>(credentialPtr);
+                return Marshal.PtrToStringUni(cred.CredentialBlob, cred.CredentialBlobSize / sizeof(char));
+            }
+            finally
+            {
+                CredFree(credentialPtr);
+            }
+        }
+
+        public static void WriteCredentials(string targetName, string secret)
+        {
+            if (Encoding.Unicode.GetByteCount(secret) > 512)
+                throw new ArgumentOutOfRangeException(nameof(secret), "The secret is greater than 512 bytes");
+
+            var credentialBlob = Marshal.StringToHGlobalUni(secret);
+            var cred = new CREDENTIAL
+            {
+                Type = CredentialTypeGeneric,
+                TargetName = targetName,
+                CredentialBlobSize = secret.Length * sizeof(char),
```

Comment:
Not sure why we use `Encoding.Unicode.GetByteCount` above, but assume it's `length * sizeof(char)` here. If it's unicode, do we need to worry about "characters" encoded with more than one Unicode code point?